### DISCUSSION TITLE
Add workspace name to  `rofi -show window` output

### DISCUSSION THREAD

**Discussion by blueray453** [davatorium/rofi#1869] [Q&A] [answered]
The `rofi -steal-focus -show window` command output has three columns. I want to add 2 more  column with the workspace name and and workspace number. By name I mean:

    gsettings set org.gnome.desktop.wm.preferences workspace-names "['Nemo', 'Firefox', 'Codium', 'Terminals', 'Books', 'Notes']"

How can I achieve that?

**Comment by DaveDavenport** [comment, marked as answer]
https://davatorium.github.io/rofi/1.7.5/rofi.1/#window-switcher-settings

If it going to work depends on your window manager exposing the right information via X11.

**Comment by blueray453** [comment]
I am using gnome shell. Debian 12.

```
$ printf 'Desktop: %s\nSession: %s\n' "$XDG_CURRENT_DESKTOP" "$GDMSESSION"
Desktop: GNOME
Session: gnome
```

**Reply by DaveDavenport** [reply]
I don't use gnome, so no idea if it will work.

**Reply by blueray453** [reply]
That's ok. Thank you very much.

**Reply by blueray453** [reply]
A quick followup.

I have checked 

https://github.com/davatorium/rofi-plugin-template
https://github.com/davatorium/rofi/blob/next/source/modes/window.c

I will not be able to write c plugins.

I can write **bash script**. 

What I can deal with is:

https://davatorium.github.io/rofi/1.7.5/rofi-script.5/

I am familiar with xdotool.

* Can get list of workspaces.
* Can know which window belong to which workspace.

What I am interested to know is: 

* how to create columns (I think we have to use pango markup `-markup-rows`). I also looked at `-display-columns` and `-display-column-separator`. not understanding how to do it so that it looks like the builtin window or windowcd mode.
* how to give icons in the first column.

If you can give me a template (with n columns and dummy data, and also icon in first column) then it will be very helpful. I am planning to write multiple scripts for gnome and hopefully this will be a easy beginning.

**Reply by DaveDavenport** [reply]
* Rofi has no real 'columns ' support,  it depends on (monospace) font and/or using spaces/tabs.
* icons, see rofi-script manpage (icon)

What is currently missing in the `window` mode?  I am unsure how we went to 'is this possible' to 'write a plugin' without remarking what does not work.

**Comment by blueray453** [comment]
If someone ask for a similar problem (using gnome), they can use the following code:

```
#!/usr/bin/env bash

shopt -s lastpipe

json=$(dbus-send --print-reply=literal --session --dest=org.gnome.Shell /org/gnome/Shell/Extensions/GnomeUtilsWorkspaces org.gnome.Shell.Extensions.GnomeUtilsWorkspaces.GetWorkspaces)

windows_array=($( echo $json | jq -c -r '.all_normal_windows_of_workspaces[][]' ))

workspace_names=($(echo $json | jq -c -r '.workspace_names[]'))

status=0

for w in ${workspace_names[@]}; do

    if [ $status -eq 0 ]; then
        status=1
        echo -en "\tWorkspaces\tWindow Titles\0nonselectable\x1ftrue\n"
    fi

    all_normal_windows_of_nth_workspaces=$( echo $json | jq ".all_normal_windows_of_workspaces .$w" | jq -c '.[]' )

    for item in ${all_normal_windows_of_nth_workspaces}; do
        title=$(dbus-send --print-reply=literal --session --dest=org.gnome.Shell /org/gnome/Shell/Extensions/GnomeUtilsWindows org.gnome.Shell.Extensions.GnomeUtilsWindows.GetTitle uint32:$item | tr -d [:blank:])

        icon=$(dbus-send --print-reply=literal --session --dest=org.gnome.Shell /org/gnome/Shell/Extensions/GnomeUtilsWindows org.gnome.Shell.Extensions.GnomeUtilsWindows.GetIconFromWinID uint32:$item | tr -d [:blank:])
        echo -en "\t${w}"
        echo -en "\t${title}"
        echo -en "\0icon\x1f${icon}\n"
        # echo -en "aap\0icon\x1f${icon}\n" | head -1 | od -c
    done

done | rofi -dmenu -format 'i' -theme-str 'element-text {tab-stops:[1em,6em];}' | read select

if [ ${PIPESTATUS[1]} -eq 0 ]; then
    dbus-send --print-reply=literal --session --dest=org.gnome.Shell /org/gnome/Shell/Extensions/GnomeUtilsWindows org.gnome.Shell.Extensions.GnomeUtilsWindows.Activate uint32:${windows_array[$(($select-1))]}
fi
``` 
For this to work they will need to use the extension https://github.com/blueray453/gnome-utils-by-blueray453

Few issues. 

I tried to use:

    if [ $status -eq 0 ]; then
        status=1
        echo -en "\0theme\x1felement-text{ tab-stops:[1em,6em];}"
        echo -en "\0markup-rows\x1ftrue"
        echo -en "\tWorkspaces\tWindow Titles\0nonselectable\x1ftrue\n"
        
    fi

But it removes the first column with an empty column. So, I used 

    if [ $status -eq 0 ]; then
        status=1
        echo -en "\tWorkspaces\tWindow Titles\0nonselectable\x1ftrue\n"
    fi

If you have any idea why this is happening then please let me know.

Those who want to use this code may need to change the tab-stops values in `rofi -dmenu -format 'i' -theme-str 'element-text {tab-stops:[1em,6em];}'`.

Here is a screenshot:

![final](https://github.com/davatorium/rofi/assets/855297/6c5fcd37-df22-4611-8ab7-011711f85253)

**Reply by DaveDavenport** [reply]
> If you have any idea why this is happening then please let me know.

You have missing newlines on the first two echos.

**Reply by blueray453** [reply]
If you mean,

    echo -en "\0theme\x1felement-text{ tab-stops:[1em,6em];}"
    echo -en "\0markup-rows\x1ftrue"

I thought as these lines will not print anything, i will not need to use `\n` there.

fixing it.

**Reply by DaveDavenport** [reply]
newline is the separator, without a separator it is all one entry.

this is what is received:
```bash
    echo -en "\0theme\x1felement-text{ tab-stops:[1em,6em];}\0markup-rows\x1ftrue\tWorkspaces\tWindow Titles\0nonselectable\x1ftrue\n"
```

**Reply by blueray453** [reply]
Thank you very much.
